src/Array.ts

```diff
@@ -741,7 +742,21 @@ export const rotate = <A>(n: number, xs: Array<A>): Array<A> => {
  */
 export const uniq = <A>(S: Setoid<A>) => (as: Array<A>): Array<A> => {
   const r: Array<A> = []
-  new Set(as).forEach(a => r.push(a))
+  const set: Set<A> = new Set()
+  const isMember = member(S)(set)
+  const l = as.length
+  let i = 0
+  while (i < l) {
+    const a = as[i]
+    if (!isMember(a)) {
+      set.add(a)
+    }
+    i += 1
+  }
+  if (l === set.size) {
+    return as
+  }
+  set.forEach(a => r.push(a))
```

Comment:
@alex-ketch you can presize the array, it will be more efficient (create less realloc and is generally faster).
Please take a look at:
https://github.com/alex-ketch/fp-ts/blob/d59e64d9a0510dbd4481c7f0460796c63de2fb91/src/Array.ts#L537-L544